rfcs/20200721-extension-types.md

```diff
@@ -0,0 +1,1304 @@
+# Extension Types
+
+| Status        | Proposed                                             |
+:-------------- |:---------------------------------------------------- |
+| **Authors**   | Edward Loper (edloper@google.com) |
+| **Sponsor**   | Alex Passos (apassos@google.com)                     |
+| **Updated**   | 2020-07-21                                           |
+
+## Objective
+
+This RFC proposes a protocol that can be used to define **_user-defined
+object-oriented Python types_** that are supported by TensorFlow APIs.
+
+## Motivation
+
+Object oriented types can make systems more readable, modular, maintainable.
+However, most TensorFlow APIs do not currently support user-defined Python
+types.  This includes both high-level APIs (such as `Keras`, `tf.function`,
+`tf.SavedModel`) and lower-level APIs (such as `tf.while_loop` and `tf.add`).  
+
+This RFC proposes a set of protocols that will allow TensorFlow APIs to handle
+user-defined Python types.  A version of this interface is already used
+internally to implement several core TensorFlow data types, including
+`tf.SparseTensor`, `tf.RaggedTensor`, `tf.data.Dataset`, and
+`tf.StructuredTensor`.
+
+At a high level, types supported by this interface can be divided into two broad
+categories:
+
+* **_General data structures_**.
+  These types are handled by "generic" APIs whose behavior does not depend on
+  the value of each object (such as `tf.function`, `SavedModel`, and
+  `tf.while_loop`).
+
+* **_Tensor-like types_**, which specialize or extend tf.Tensor.
+  Values of these types have a `rank`, a `shape`, and usually a `dtype`.  In
+  addition to the "generic" APIs, these types can be handled by Tensor-specific
+  APIs (such as `tf.stack`, `tf.add`, and `tf.reduce_mean`).
+
+Examples of user-defined types that could defined or extended with this protocol
+include:
+
+**General data structures:**
+
+* `tfp.Distribution`: Encodes a statistical distribution.
+* `TensorDigraph`: Encodes the set of nodes and edges in a directed graph.
+* `DimensionAlignment`: Encodes a correspondence between two related dimensions
+  (e.g., between a `word` dimension and a `speaker` dimension).
+
+**Tensor-like types:**
```

Comment:
Another obvious example would be some form of distributed tensor, like those from Mesh-TensorFlow.